src/libs/elektra/keyset.c

```diff
@@ -1099,12 +1172,15 @@ ssize_t ksRename (KeySet * ks, const Key * root, const Key * newRoot)
 	if (ks == NULL || root == NULL || newRoot == NULL) return -1;
 	if (keyGetNamespace (root) == KEY_NS_CASCADING || keyGetNamespace (newRoot) == KEY_NS_CASCADING) return -1;
 
+	// TODO: do we really need to detach here?
+	keySetDetachData (ks);
```

Comment:
Yes, we need that. The function changes the keys contained in this keyset **without** changing them in other keysets they might be in. That's why `ksRenameInternal` has the "dup in-place".

This function is a bit weird and could be explained better. What it does is essentially the following, but much more optimized.

```c
elektraCursor end;
for (elektraCursor it = ksFindHierarchy (ks, root, &end); it < end; it++) {
    Key * k = elektraKsPopAtCursor (ks, it);
    Key * dup = keyDup (k); // needed because KEY_FLAG_RO_NAME
    keyDel (k);
    keyReplacePrefix (dup, root, newRoot);
    ksAppendKey (ks, dup);
}
```

But with all the COW, there may now be a better way to achieve this.